File: data_structures/linked_list/singly_linked_list.js
/**
 * Singly Linked List implementation in JavaScript
 */
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.size = 0;
    }

    isEmpty() {
        return this.head === null;
    }

    length() {
        return this.size;
    }

    insertHead(data) {
        let newNode = new Node(data);
        newNode.next = this.head;
        this.head = newNode;
        this.size++;
    }

    insertTail(data) {
        let newNode = new Node(data);
        if (this.head === null) {
            this.head = newNode;
        } else {
            let current = this.head;
            while (current.next !== null) {
                current = current.next;
            }
            current.next = newNode;
        }
        this.size++;
    }

    insertNth(index, data) {
        if (index < 0 || index > this.size) return;
        if (index === 0) {
            this.insertHead(data);
            return;
        }
        let newNode = new Node(data);
        let current = this.head;
        for (let i = 0; i < index - 1; i++) {
            current = current.next;
        }
        newNode.next = current.next;
        current.next = newNode;
        this.size++;
    }

    deleteHead() {
        if (this.head === null) return null;
        let data = this.head.data;
        this.head = this.head.next;
        this.size--;
        return data;
    }

    deleteTail() {
        if (this.head === null) return null;
        if (this.head.next === null) return this.deleteHead();

        let current = this.head;
        while (current.next.next !== null) {
            current = current.next;
        }
        let data = current.next.data;
        current.next = null;
        this.size--;
        return data;
    }

    reverse() {
        let prev = null;
        let current = this.head;
        while (current !== null) {
            let nextNode = current.next;
            current.next = prev;
            prev = current;
            current = nextNode;
        }
        this.head = prev;
    }

    printList() {
        let current = this.head;
        let result = [];
        while (current !== null) {
            result.push(current.data);
            current = current.next;
        }
        console.log(result.join(" -> "));
    }
}

const ll = new LinkedList();
ll.insertTail(1);
ll.insertTail(2);
ll.insertTail(3);
ll.printList();

ll.insertHead(0);
ll.printList();

ll.reverse();
ll.printList();

console.log("Deleted head:", ll.deleteHead());
ll.printList();

File: data_structures/linked_list/singly_linked_list.cpp
#include <iostream>
#include <string>
#include <vector>

// Generic Node using templates
template <typename T> struct Node {
  T data;
  Node *next;

  Node(T val) : data(val), next(nullptr) {}
};

template <typename T> class LinkedList {
private:
  Node<T> *head;
  int size;

public:
  LinkedList() : head(nullptr), size(0) {}

  // Destructor to clean up memory
  ~LinkedList() {
    Node<T> *current = head;
    while (current != nullptr) {
      Node<T> *next = current->next;
      delete current;
      current = next;
    }
  }

  bool isEmpty() const { return head == nullptr; }

  int length() const { return size; }

  void insertHead(T data) {
    Node<T> *newNode = new Node<T>(data);
    newNode->next = head;
    head = newNode;
    size++;
  }

  void insertTail(T data) {
    Node<T> *newNode = new Node<T>(data);
    if (head == nullptr) {
      head = newNode;
    } else {
      Node<T> *temp = head;
      while (temp->next != nullptr) {
        temp = temp->next;
      }
      temp->next = newNode;
    }
    size++;
  }

  void insertNth(int index, T data) {
    if (index < 0 || index > size) {
      std::cerr << "Index out of range" << std::endl;
      return;
    }
    if (index == 0) {
      insertHead(data);
      return;
    }
    Node<T> *newNode = new Node<T>(data);
    Node<T> *temp = head;
    for (int i = 0; i < index - 1; ++i) {
      temp = temp->next;
    }
    newNode->next = temp->next;
    temp->next = newNode;
    size++;
  }

  T deleteHead() {
    if (head == nullptr) {
      throw std::runtime_error("List is empty");
    }
    Node<T> *temp = head;
    T data = temp->data;
    head = head->next;
    delete temp;
    size--;
    return data;
  }

  T deleteTail() {
    if (head == nullptr) {
      throw std::runtime_error("List is empty");
    }
    if (head->next == nullptr) {
      return deleteHead();
    }
    Node<T> *temp = head;
    while (temp->next->next != nullptr) {
      temp = temp->next;
    }
    Node<T> *toDelete = temp->next;
    T data = toDelete->data;
    temp->next = nullptr;
    delete toDelete;
    size--;
    return data;
  }

  void reverse() {
    Node<T> *prev = nullptr;
    Node<T> *current = head;
    while (current != nullptr) {
      Node<T> *nextNode = current->next;
      current->next = prev;
      prev = current;
      current = nextNode;
    }
    head = prev;
  }

  void printList() const {
    Node<T> *temp = head;
    while (temp != nullptr) {
      std::cout << temp->data;
      if (temp->next != nullptr)
        std::cout << " -> ";
      temp = temp->next;
    }
    std::cout << std::endl;
  }
};

int main() {
  LinkedList<int> ll;
  ll.insertTail(1);
  ll.insertTail(2);
  ll.insertTail(3);
  ll.printList(); // 1 -> 2 -> 3

  ll.insertHead(0);
  ll.printList(); // 0 -> 1 -> 2 -> 3

  ll.reverse();
  ll.printList(); // 3 -> 2 -> 1 -> 0

  std::cout << "Deleted head: " << ll.deleteHead() << std::endl;
  ll.printList(); // 2 -> 1 -> 0

  return 0;
}
